--- a/make/Makefile.linux-kmodule
+++ b/make/Makefile.linux-kmodule
@@ -71,6 +71,10 @@
 # automatically got all these flags.
 #
 ccflags-y += $(CFGFLAGS) $(LKM_BUILD_INFO)
+ccflags-y += -UKCOM_FILTER_MAX -DKCOM_FILTER_MAX=1025
+ccflags-y += -UKCOM_NETIF_MAX -DKCOM_NETIF_MAX=1056
+ccflags-y += -UNGKNET_NETIF_MAX -DNGKNET_NETIF_MAX=1056
+ccflags-y += -UNGKNET_FILTER_MAX -DNGKNET_FILTER_MAX=1025
 
 # Required for older kernels
 export EXTRA_CFLAGS := $(ccflags-y)
